core/primitives/src/types.rs

```diff
@@ -401,17 +390,275 @@ pub struct ApprovalStake {
+            Self::V1(ValidatorStakeV1 { account_id, public_key, stake })
+        }
+
+        pub fn into_v1(self) -> ValidatorStakeV1 {
+            match self {
+                Self::V1(v1) => v1,
+            }
+        }
+
+        #[inline]
+        pub fn account_and_stake(self) -> (AccountId, Balance) {
+            match self {
+                Self::V1(v1) => (v1.account_id, v1.stake),
+            }
+        }
+
+        #[inline]
+        pub fn destructure(self) -> (AccountId, PublicKey, Balance) {
+            match self {
+                Self::V1(v1) => (v1.account_id, v1.public_key, v1.stake),
+            }
+        }
+
+        #[inline]
+        pub fn take_account_id(self) -> AccountId {
+            match self {
+                Self::V1(v1) => v1.account_id,
+            }
+        }
+
+        #[inline]
+        pub fn account_id(&self) -> &AccountId {
+            match self {
+                Self::V1(v1) => &v1.account_id,
+            }
+        }
+
+        #[inline]
+        pub fn take_public_key(self) -> PublicKey {
+            match self {
+                Self::V1(v1) => v1.public_key,
+            }
+        }
+
+        #[inline]
+        pub fn public_key(&self) -> &PublicKey {
+            match self {
+                Self::V1(v1) => &v1.public_key,
+            }
+        }
+
+        #[inline]
+        pub fn stake(&self) -> Balance {
+            match self {
+                Self::V1(v1) => v1.stake,
+            }
+        }
+
+        #[inline]
+        pub fn stake_mut(&mut self) -> &mut Balance {
+            match self {
+                Self::V1(v1) => &mut v1.stake,
+            }
+        }
+
+        pub fn get_approval_stake(&self, is_next_epoch: bool) -> ApprovalStake {
+            ApprovalStake {
+                account_id: self.account_id().clone(),
+                public_key: self.public_key().clone(),
+                stake_this_epoch: if is_next_epoch { 0 } else { self.stake() },
+                stake_next_epoch: if is_next_epoch { self.stake() } else { 0 },
+            }
+        }
+    }
+}
+
+#[cfg(not(feature = "protocol_feature_block_header_v3"))]
+pub mod validator_stake {
+    use crate::types::ApprovalStake;
+    use borsh::{BorshDeserialize, BorshSerialize};
+    use near_crypto::PublicKey;
+    use near_primitives_core::types::{AccountId, Balance};
+    use serde::Serialize;
+
+    /// Stores validator and its stake.
+    #[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+    pub struct ValidatorStake {
+        /// Account that stakes money.
+        pub account_id: AccountId,
+        /// Public key of the proposed validator.
+        pub public_key: PublicKey,
+        /// Stake / weight of the validator.
+        pub stake: Balance,
+    }
+
+    pub type ValidatorStakeV1 = ValidatorStake;
+
+    pub struct ValidatorStakeIter<'a> {
+        collection: &'a [ValidatorStake],
+        curr_index: usize,
+        len: usize,
+    }
```

Comment:
I think can define ValidatorStakeIter as `struct ValidatorStakeIter<'a>(std::slice::iter<'a, ValidatorStake>), so don't need keep track of curr_index and len and reuse https://doc.rust-lang.org/std/slice/struct.Iter.html#impl-Iterator to impl Iterator

Reply:
Yes, reusing the iterator is a good idea. Done in https://github.com/near/nearcore/pull/4115/commits/14273cdea40d52322e38656b4be88ce9f1245caa

I kept the `len` field though because there are places in the code where knowing the number of elements without consuming the iterator is important.